### ISSUE TITLE
Unnecessary fences in HeaderWriter

### ISSUE THREAD

**Issue by buybackoff** [AdaptiveConsulting/Aeron.NET#29]
Unless I am missing something inside MediaDriver implementation, it looks like both fences are not needed [here](https://github.com/AdaptiveConsulting/Aeron/blob/master/src/Adaptive.Aeron/LogBuffer/HeaderWriter.cs#L44) because a frame is committed only after its lengths is written atomically. All usages of `headerWriter.Write` are followed by `FrameDescriptor.FrameLengthOrdered`, which is implemented as `Volatile.Write` that [implicitly generates](http://stackoverflow.com/questions/6581848/memory-barrier-generators) a memory barrier. TermReader then reads that length through Volatile.Read [here](https://github.com/AdaptiveConsulting/Aeron/blob/master/src/Adaptive.Aeron/LogBuffer/TermReader.cs#L37) and doesn't touch any other frame part until the length is positive. On my machine replacing the two lines with a simple `PutLong` increases IPC performance from c.14 Mops to c.16 Mops.

In .NET we could even create a [DataHeader](https://github.com/Spreads/Spreads/blob/f01ddbd889b0f09e776cd689cf38435bc23315f9/src/Spreads.Extensions/Storage/Aeron/Protocol/DataHeaderFlyweight.cs#L9) struct and write it directly like [this](https://github.com/Spreads/Spreads/blob/f01ddbd889b0f09e776cd689cf38435bc23315f9/src/Spreads.Extensions/Storage/Aeron/Logbuffer/HeaderWriter.cs#L35). Also, HeaderWriter could be made sealed or a struct. These additional changes together add c.0.5 Mops to the IPCThroughput sample.

**Comment by mjpt777** [comment]
The fence ordering is important to detection of dead clients that have written a partial message. The driver can unblock streams when a client crashes but to do so it follows a protocol that the client must honour.

**Comment by buybackoff** [comment]
@mjpt777 Thanks, I understand now that after position atomic increment a client could die without committing. Could you please give a link to the driver Java code line where this detection is happening? My experience with Java and IDEA is quite limited, couldn't find this line before.

**Comment by buybackoff** [comment]
In .NET, only one fence is enough. There is a comment with TODO about that by James. Volatile generates the needed fence already.

**Comment by mjpt777** [comment]
https://github.com/real-logic/Aeron/blob/master/aeron-client/src/main/java/io/aeron/logbuffer/TermUnblocker.java

You need a StoreStore AFTER the atomic write of the negative length that comes before the write of the message body in bytes or other header fields. Unsafe.putOrderedX in Java has a StoreStore before the atomic write therefore the storeFence() is required after. There are a lot of subtleties to the ordering hear and best to be conservative unless you are absolutely sure what you are doing.

**Comment by buybackoff** [comment]
@mjpt777 What will happen if a writer dies after incrementing the tail  but before reaching [this line](https://github.com/real-logic/Aeron/blob/master/aeron-client/src/main/java/io/aeron/logbuffer/HeaderWriter.java#L86)? In this scenario the negative length is not written but the tail is already incremented - the unblocker won't be able to do anything.

With the changes below the tail is incremented only after a negative length value is written atomically. 
* If a client dies right after writing negative length but before tail increment, conductor will see the negative value and could unblock properly.
* If a client dies after tail increment but before committing, current logic applies.
* TermAppender could detect blocked term and unblock it itself, without conductor help.

Here is [Claim method implementation](https://github.com/Spreads/Spreads/blob/master/src/Spreads.Extensions/Storage/Aeron/Logbuffer/TermAppender.cs#L91) in my version.


Below are TermAppender.AppendUnfragmentedMessage method and HeaderWriter class . On my machine IPC performance increases from c.14 to c.15.3 Mops. Note that I use direct Interlocked and Volatile methods on buffer pointers - bound checks are not needed here (already done via `resultingOffset > termLength`) and also we avoid interface method call as discussed in #33.
 
```
        public unsafe long AppendUnfragmentedMessage(HeaderWriter header, IDirectBuffer srcBuffer, int srcOffset, int length) {
            int frameLength = length + DataHeaderFlyweight.HEADER_LENGTH;
            int alignedLength = BitUtil.Align(frameLength, FrameDescriptor.FRAME_ALIGNMENT);

            IAtomicBuffer termBuffer = _termBuffer;
            int termLength = termBuffer.Capacity;
            long resultingOffset;
            var spinCounter = 0;
            var rawTail = RawTailVolatile();

            while (true) {
                
                var termOffset = rawTail & 0xFFFFFFFFL;
                resultingOffset = termOffset + alignedLength;

                if (resultingOffset > termLength) {
                    // NB: avoid method call, write directly: _metaDataBuffer.PutLongOrdered(LogBufferDescriptor.TERM_TAIL_COUNTER_OFFSET, rawTail + alignedLength);
                    Volatile.Write(ref *(long*)(_metaDataBuffer.BufferPointer + LogBufferDescriptor.TERM_TAIL_COUNTER_OFFSET), rawTail + alignedLength);
                    resultingOffset = HandleEndOfLogCondition(termBuffer, termOffset, header, termLength, TermId(rawTail));
                    break;
                }

                // true if we are the first to claim space at current offset
                if (0 ==
                    Interlocked.CompareExchange(
                        ref *(int*)(new IntPtr(_termBuffer.BufferPointer.ToInt64() + termOffset)), -length, 0)) {
                    // if a writer dies here, another writer will unblock below
                    // NB: avoid method call, write directly: _metaDataBuffer.PutLongOrdered(LogBufferDescriptor.TERM_TAIL_COUNTER_OFFSET, rawTail + alignedLength);
                    Volatile.Write(ref *(long*) (_metaDataBuffer.BufferPointer + LogBufferDescriptor.TERM_TAIL_COUNTER_OFFSET), rawTail + alignedLength);
                    int offset = (int)termOffset;
                    header.Write(termBuffer, offset, frameLength, TermId(rawTail));
                    termBuffer.PutBytes(offset + DataHeaderFlyweight.HEADER_LENGTH, srcBuffer, srcOffset, length);
                    FrameDescriptor.FrameLengthOrdered(termBuffer, offset, frameLength);
                    break;
                }

                // spin, will re-read (volatile) current tail and try again
                // single writer will always succeed on first try
                var previousRawTail = rawTail;
                rawTail = RawTailVolatile();
                if (previousRawTail == rawTail) {
                    // incrementing tail happens right next to interlocked -length write
                    // we should spin in case another writer has written -length but not yet incremented tail
                    spinCounter++;
                    if (spinCounter > 100) {
                        // no-one is progressing, need to unblock
                        // NB: avoid method call in _termBuffer.PutIntOrdered((int)termOffset, 0), write directly;
                        Volatile.Write(ref *(int*)(new IntPtr(_termBuffer.BufferPointer.ToInt64() + termOffset)), 0);
                        
                    }
                }
            }

            return resultingOffset;
        }

    public class HeaderWriter
    {
        private readonly int _versionFlagsType;
        private readonly long _sessionId;
        private readonly long _streamId;

        public HeaderWriter()
        {
        }

        public HeaderWriter(IDirectBuffer defaultHeader)
        {
            _versionFlagsType = defaultHeader.GetInt(HeaderFlyweight.VERSION_FIELD_OFFSET);
            _sessionId = (long)defaultHeader.GetInt(DataHeaderFlyweight.SESSION_ID_FIELD_OFFSET) << 32;
            _streamId = defaultHeader.GetInt(DataHeaderFlyweight.STREAM_ID_FIELD_OFFSET) & 0xFFFFFFFFL;
        }

        public virtual void Write(IAtomicBuffer termBuffer, int offset, int length, int termId)
        {
            var termOffsetSessionId = _sessionId | (uint)offset;
            var streamAndTermIds = _streamId | ((long)termId << 32);

            termBuffer.PutInt(offset + HeaderFlyweight.VERSION_FIELD_OFFSET, _versionFlagsType);
            termBuffer.PutLong(offset + DataHeaderFlyweight.TERM_OFFSET_FIELD_OFFSET, termOffsetSessionId);
            termBuffer.PutLong(offset + DataHeaderFlyweight.STREAM_ID_FIELD_OFFSET, streamAndTermIds);
        }
    }

```

I wanted to create a pull request with those changes, but my fork disappeared, so I cannot continue any further experiments until the project is released publicly.

**Comment by mjpt777** [comment]
Incrementing the tail and not making further progress is a valid step. The unblocker can detect each potential step and adjust its search accordingly. This is detected by the tail having advanced and the length being zero after the timeout. Next state is negative length set. Then last step is length flipped from negative to positive.

When the length is zero after a time out then a search forward is performed from current position to tail in alignment steps. Another message could start before tail which needs to be detected.

The performance differences to Java are so significant that I don't believe this is the main cause. It is more likely to be lack of inlining and other compiler optimisations ganging up. Best approach would be to compare the ASM generated for both Java and .NET and then work out how to bridge the gap. The gap is approximately 30-40ns per operation and a fence is only 12 cycles on the latest processors. There is a bigger win to find.

**Comment by buybackoff** [comment]
@mjpt777 What numbers you have for Java and which benchmarks you use? I am running [EmbeddedIpcThroughput](https://github.com/real-logic/Aeron/blob/master/aeron-samples/src/main/java/io/aeron/samples/EmbeddedIpcThroughput.java) from IDEA (version 0.9.7) and get only 10 Mops.

**Comment by mjpt777** [comment]
For that benchmark I get ~30m ops in Java and 32-33m in C++. I'm on a 4 core machine with hyperthreading so 8 effective threads.

**Comment by JPWatson** [comment]
@buybackoff if you run it outside IDEA, it goes much faster

I've written a bat file in the aeron-samples/scripts folder than runs it:
```
@echo off
"%JAVA_HOME%\bin\java" ^
    -cp ..\build\libs\samples.jar ^
    -XX:+UnlockDiagnosticVMOptions ^
    -XX:GuaranteedSafepointInterval=300000 ^
    -Daeron.sample.messageLength=32 ^
    -Dagrona.disable.bounds.checks=true ^
    io.aeron.samples.EmbeddedIpcThroughput
```

**Comment by buybackoff** [comment]
It is c.23.5 on my machine.
